glommio/src/net/tcp_socket.rs

```diff
@@ -0,0 +1,1022 @@
+use futures_lite::io::{AsyncBufRead, AsyncRead, AsyncWrite};
+use futures_lite::ready;
+use futures_lite::stream::{self, Stream};
+use iou::{InetAddr, SockAddr};
+use nix::sys::socket::{self, sockopt::ReusePort};
+use pin_project_lite::pin_project;
+use socket2::{Domain, Protocol, Socket, Type};
+use std::io;
+use std::net::{self, Shutdown, SocketAddr, ToSocketAddrs};
+use std::os::unix::io::RawFd;
+use std::os::unix::io::{AsRawFd, FromRawFd};
+use std::pin::Pin;
+use std::rc::Rc;
+use std::task::{Context, Poll};
+
+#[derive(Debug)]
+/// A TCP socket server, listening for connections.
+///
+/// After creating a TcpListener by binding it to a socket address, it listens for incoming TCP connections.
+/// These can be accepted by calling [`accept`] or [`shared_accept`], or by iterating over the Incoming iterator returned by [`incoming`].
+///
+/// A good networking architecture within a thread-per-core model needs to take into account
+/// parallelism and spawn work into multiple executors. If everything happens inside the same
+/// Executor, then at most one thread is used. Sometimes this is what you want: you may want to
+/// dedicate a CPU entirely for networking, or even use specialized ports for each CPU of the
+/// application, but most likely it isn't.
+///
+/// There are two approaches to load balancing possible with the `TcpListener`:
+///
+/// * By default, the ReusePort flag is set in the socket automatically. The OS already provides
+///   some load balancing capabilities with that so you can simply [`bind`] to the same address
+///   from many executors.
+///
+/// * If that is insufficient or otherwise not desirable, it is possible to use [`shared_accept`]
+///   instead of [`accept`]: that returns an object that implements [`Send`]. You can then use a
+///   [`shared_channel`] to send the accepted connection into multiple executors. The object
+///   returned by [`shared_accept`] can then be bound to its executor with [`bind_to_executor`], at
+///   which point it becomes a standard [`TcpStream`].
+///
+/// Relying on the OS is definitely simpler, but which approach is better depends on the specific
+/// needs of your application.
+///
+/// The socket will be closed when the value is dropped.
+///
+/// [`accept`]: TcpListener::accept
+/// [`shared_accept`]: TcpListener::shared_accept
+/// [`bind`]: TcpListener::bind
+/// [`incoming`]: TcpListener::incoming
+/// [`bind_to_executor`]: AcceptedTcpStream::bind_to_executor
+/// [`Send`]: https://doc.rust-lang.org/std/marker/trait.Send.html
+/// [`shared_channel`]: ../channels/shared_channel/index.html
+pub struct TcpListener {
+    reactor: Rc<Reactor>,
+    listener: net::TcpListener,
+}
+
+impl TcpListener {
+    /// Creates a TCP listener bound to the specified address.
+    ///
+    /// Binding with port number 0 will request an available port from the OS.
+    ///
+    /// This method sets the ReusePort option in the bound socket, so it is designed to
```

Comment:
I thought ReusePort was only valid within the same process space?
As an example point my tests were all using the same port for convenience in the beginning and they would fail if ran in parallel.

Reply:
Oh maybe I'm thinking of ReuseAddr.. or maybe just completely off. Lemme look into it and I'll re-raise if needed.